controllers/utils/multicloud_util.go

```diff
@@ -0,0 +1,330 @@
+		// Create or update the managedClusterAction to approve the install plan.
+		mcaName := GetManagedClusterObjectName(
+			ManagedClusterActionPrefix, clusterGroupUpgrade, "InstallPlan", installPlan.Name)
+		_, err := EnsureManagedClusterActionForInstallPlan(ctx, c, mcaName, clusterName, installPlan, Log)
+		if err != nil {
+			return InstallPlanCannotBeApproved, err
+		}
+
+		// Check that the ManagedClusterAction completed successfully. If so, then there is no need to also check
+		// the install plan.
+		time.Sleep(ManagedClusterActionWaitTimeSec * time.Second)
+		err = CheckManagedClusterActionDone(ctx, c, mcaName, clusterName, Log)
+		if err != nil {
+			Log.Info("There was an issue with the ManagedClusterAction for InstallPlan", "error", err.Error())
+			return InstallPlanCannotBeApproved, err
+		}
+		return InstallPlanWasApproved, nil
+	}
+
+	return InstallPlanCannotBeApproved, nil
+}
+
+func EnsureManagedClusterView(
+	ctx context.Context, c client.Client, name string, namespace string, resourceType string,
+	resourceName string, resourceNamespace string, cguLabel string, Log logr.Logger) (*viewv1beta1.ManagedClusterView, error) {
+
+	mcv := &viewv1beta1.ManagedClusterView{}
+	err := c.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, mcv)
+
+	if err != nil {
+		// If the specific managedClusterView was not found, create it.
+		if errors.IsNotFound(err) {
+			Log.Info("[EnsureManagedClusterView] MCV doesn't exist, create it", "name", name, "namespace", namespace)
+			viewMeta := metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+				Labels: map[string]string{
+					"openshift-cluster-group-upgrades/clusterGroupUpgrade": cguLabel,
+				},
+			}
+			viewSpec := viewv1beta1.ViewSpec{
+				Scope: viewv1beta1.ViewScope{
+					Resource:  resourceType,
+					Name:      resourceName,
+					Namespace: resourceNamespace,
+				},
+			}
+			mcv = &viewv1beta1.ManagedClusterView{
+				ObjectMeta: viewMeta,
+				Spec:       viewSpec,
+			}
+
+			if err := c.Create(ctx, mcv); err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	} else {
+		// If the specific managedClusterView was found, update it.
+		Log.Info("[EnsureManagedClusterView] MCV already exists, update it", "name", name, "namespace", namespace)
+		// Make sure labels contain the referral to the CGU.
+		labels := mcv.GetLabels()
+		if labels == nil {
+			labels = make(map[string]string)
+		}
+		labels["openshift-cluster-group-upgrades/clusterGroupUpgrade"] = cguLabel
+		mcv.SetLabels(labels)
+
+		viewSpec := viewv1beta1.ViewSpec{
+			Scope: viewv1beta1.ViewScope{
+				Resource:  resourceType,
+				Name:      resourceName,
+				Namespace: resourceNamespace,
+			},
+		}
+
+		mcv.Spec = viewSpec
+		if err := c.Update(ctx, mcv); err != nil {
+			return nil, err
+		}
+	}
+	// Give some time for the ManagedClusterView to retrieve the info.
+	time.Sleep(ManagedClusterActionWaitTimeSec * time.Second)
```

Comment:
Yes, I agree, however after consulting with ACM folks, it felt like the better option. It's either we wait a couple of seconds now and have a really good chance of getting the info we need or wait 5 minutes until the next reconcile loop. It can save quite a lot of time in the end doing it this way.

> Irina Mihai Jan 20th at 16:42
> Hi all! What's the average time we should expect a managedClusterAction CR to be Completed? I've noticed it's usually pretty fast, but was wondering what's the amount of time we should wait before checking its status.
> Thank you!
> 2 replies
> Dangpeng Liu:no_entry:  7 days ago
> Actually, there is no average time for that. We just watch the managedClusterAction CR and apply the resource in the managedClusterAction template. So the response time may be affected by managedCluster api server.
> A better way is check the status condition Completed periodically (interval could be 1-3 seconds). And set an timeout (may be 1-2 minute) (edited)

Reply:
This is a case where an exponential backoff loop for the reconcile time would be helpful. On the first pass the loop would re-trigger after 5 seconds, then 10, then 20, up to the max limit.